ctru-rs/src/services/uds.rs

```diff
@@ -0,0 +1,1273 @@
+    /// use ctru::services::uds::{ConnectionType, Uds};
+    /// let mut uds = Uds::new(None)?;
+    ///
+    /// let networks = uds.scan(b"HBW\x10", None, None)?;
+    /// uds.connect_network(&networks[0], b"udsdemo passphrase c186093cd2652741\0", ConnectionType::Client, 1)?;
+    /// if uds.wait_status_event(false, false)? {
+    ///     println!("Connection status event signalled");
+    ///     let status = uds.get_connection_status()?;
+    ///     println!("Status: {status:#X?}");
+    /// }
+    /// #
+    /// # Ok(())
+    /// # }
+    /// ```
+    #[doc(alias = "udsGetConnectionStatus")]
+    pub fn get_connection_status(&self) -> crate::Result<ConnectionStatus> {
+        let mut status = MaybeUninit::uninit();
+
+        ResultCode(unsafe { ctru_sys::udsGetConnectionStatus(status.as_mut_ptr()) })?;
+
+        let status = unsafe { status.assume_init() };
+
+        Ok(status.into())
+    }
+
+    /// Send a packet to the network.
+    ///
+    /// TODO: max size?
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the service is currently neither connected to nor hosting a network.
+    /// See [`Uds::connect_network()`] to connect to a network or [`Uds::create_network()`] to create one.
+    /// It will also return an error if the service is currently connected to a network as a spectator, as spectators cannot send data, only receive it.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # let _runner = test_runner::GdbRunner::default();
+    /// # use std::error::Error;
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// #
+    /// use ctru::services::uds::{ConnectionType, NodeID, SendFlags, Uds};
+    /// let mut uds = Uds::new(None)?;
+    ///
+    /// let networks = uds.scan(b"HBW\x10", None, None)?;
+    /// uds.connect_network(&networks[0], b"udsdemo passphrase c186093cd2652741\0", ConnectionType::Client, 1)?;
+    /// uds.send_packet(b"Hello, World!", NodeID::Broadcast, 1, SendFlags::Default)?;
+    /// #
+    /// # Ok(())
+    /// # }
+    /// ```
+    #[doc(alias = "udsSendTo")]
+    pub fn send_packet(
+        &self,
+        packet: &[u8],
+        address: NodeID,
+        channel: u8,
+        flags: SendFlags,
+    ) -> Result<(), Error> {
+        if self.service_status() == ServiceStatus::Disconnected {
+            return Err(Error::NotConnected);
+        }
+
+        if self.context.unwrap().spectator {
+            return Err(Error::Spectator);
+        }
+
+        let code = ResultCode(unsafe {
+            ctru_sys::udsSendTo(
+                address.into(),
+                channel,
+                flags.bits(),
+                packet.as_ptr().cast(),
+                packet.len(),
+            )
+        });
+
+        if code.0
+            != ctru_sys::MAKERESULT(
+                ctru_sys::RL_STATUS as _,
+                ctru_sys::RS_OUTOFRESOURCE as _,
+                ctru_sys::RM_UDS as _,
+                ctru_sys::RD_BUSY as _,
+            )
```

Comment:
What does this status represent? it still seems like it would be an error of some kind but we
return `Ok` for this?

Reply:
See [libctru](https://github.com/devkitPro/libctru/blob/master/libctru/include/3ds/services/uds.h#L23) and [the UDS example](https://github.com/devkitPro/3ds-examples/blob/master/network/uds/source/uds.c#L286C7-L286C34). There's probably a nicer way to handle this, but this does work.

Reply:
Ok, so this error code basically means a packet was still sent but some other error occurred, I guess? Seems fine

Reply:
Yeah, something like that. Maybe it'd be a good idea to hardcode it as a constant somewhere rather than calling it in the function, though - `MAKERESULT` presumably isn't `const` so it'd have to be a magic number with some explanation.

Reply:
If it isn't `const`, we probably can/should make it `const`, same for any of those other helper functions that just do bit operations on `Result`

Reply:
Presumably, this would need to be changed in ctru_sys. Once somebody does that, I'll change this to be a little more elegant.